packages/flutter/lib/src/rendering/slivers.dart

```diff
@@ -0,0 +1,1850 @@
+  }) {
+    switch (axis) {
+      case Axis.horizontal:
+        return new BoxConstraints(
+          minHeight: crossAxisExtent,
+          maxHeight: crossAxisExtent,
+          minWidth: minExtent,
+          maxWidth: maxExtent,
+        );
+      case Axis.vertical:
+        return new BoxConstraints(
+          minWidth: crossAxisExtent,
+          maxWidth: crossAxisExtent,
+          minHeight: minExtent,
+          maxHeight: maxExtent,
+        );
+    }
+    return null;
+  }
+
+  @override
+  bool debugAssertIsValid({
+    bool isAppliedConstraint: false,
+    InformationCollector informationCollector
+  }) {
+    // TODO(ianh): make these show pretty errors
+    assert(axis != null);
+    assert(growthDirection != null);
+    assert(scrollOffset != null);
+    assert(overlap != null);
+    assert(remainingPaintExtent != null);
+    assert(crossAxisExtent != null);
+    assert(scrollOffset >= 0.0);
+    assert(crossAxisExtent >= 0.0);
+    assert(remainingPaintExtent >= 0.0);
+    assert(isNormalized); // should be redundant with earlier checks
+    return true;
+  }
+
+  @override
+  bool operator ==(dynamic other) {
+    assert(debugAssertIsValid());
+    if (identical(this, other))
+      return true;
+    if (other is! SliverConstraints)
+      return false;
+    final SliverConstraints typedOther = other;
+    assert(typedOther.debugAssertIsValid());
+    return axis == typedOther.axis &&
+           growthDirection == typedOther.growthDirection &&
+           scrollOffset == typedOther.scrollOffset &&
+           overlap == typedOther.overlap &&
+           remainingPaintExtent == typedOther.remainingPaintExtent &&
+           crossAxisExtent == typedOther.crossAxisExtent;
+  }
+
+  @override
+  int get hashCode {
+    assert(debugAssertIsValid());
+    return hashValues(axis, growthDirection, scrollOffset, overlap, remainingPaintExtent, crossAxisExtent);
+  }
+
+  @override
+  String toString() {
+    return 'SliverConstraints('
+             '$axisDirection, '
+             '$growthDirection, '
+             '$userScrollDirection, '
+             'scrollOffset: ${scrollOffset.toStringAsFixed(1)}, '
+             'remainingPaintExtent: ${remainingPaintExtent.toStringAsFixed(1)}, ' +
+             (overlap != 0.0 ? 'overlap: ${overlap.toStringAsFixed(1)}, ' : '') +
+             'crossAxisExtent: ${crossAxisExtent.toStringAsFixed(1)}'
+           ')';
+  }
+}
+
+class SliverGeometry {
+  SliverGeometry({
+    this.scrollExtent: 0.0,
+    this.paintExtent: 0.0,
+    double layoutExtent,
+    double opaqueExtent,
+    this.maxPaintExtent: 0.0,
+    double hitTestExtent,
+    bool visible,
+    this.scrollOffsetCorrection: 0.0
+  }) : opaqueExtent = opaqueExtent ?? paintExtent,
+       layoutExtent = layoutExtent ?? paintExtent,
+       hitTestExtent = hitTestExtent ?? paintExtent,
+       visible = visible ?? paintExtent > 0.0 {
+    assert(debugAssertIsValid);
+  }
+
+  static final SliverGeometry zero = new SliverGeometry();
```

Comment:
I've removed the assert from the constructor and made the main constructor const.